---
title: "Maize-Teosinte limma DE"
author: "Caryn Johansen"
date: "9/25/2017"
output: html_document
---

# Introduction


I used RNA-seq data from Lemmon et al. 2014 (http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1004745), downloaded from the SRA (https://www.ncbi.nlm.nih.gov/bioproject/PRJNA262181)

```
fastq-dump --outdir data/raw --gzip --skip-technical --readids --read-filter pass --dumpbase --clip -fasta $srr
```

Index was created from the AGPv4 full genome cdna (ftp://ftp.ensemblgenomes.org/pub/plants/release-37/fasta/zea_mays/cdna/)

```
kallisto index -i $idx $cdna
```

**all details for scripts found in the project folder:** `Projects/Highland_GeneticIncompatibility/TeoMaizeCrossTest/scripts` 

Kallisto code found `scripts` under `kallisto.sh`. The general command was:

```
kallisto quant -i $idx -o data/bootstrap100/$srr -b 100 --single -l 101 -s 20 data/raw/$fastq
```

For the analysis for differential expression, I am using limma

package found at: http://bioconductor.org/packages/release/bioc/html/limma.html

# Setup


```{r}
library(limma)
library(tidyverse)
library(reshape2)
```

# Data

SRA samples:

```{r}
sample_id <- dir(file.path("data/fullTranscript/"))
sample_id
```

Get the file path for kallisto output

```{r}
kal_results <- file.path("data/fullTranscript", sample_id)
kal_results

res <- as.data.frame(cbind(sample_id, kal_results))
res
```

Generate the data structure to provide sleuth with the design matrix.

There are two tissue types here, but I really only want to do leaf until I have a reason to do both tissues.

```{r}
meta <- read.table("simple_meta.txt", header=T, sep = " ", stringsAsFactors = T)
meta_leaf <- dplyr::filter(meta, tissue_s == "leaf")
meta_leaf <- dplyr::select(meta_leaf, sample=Run_s, cultivar=cultivar_s)
str(meta_leaf)

sample_leaf <- as.character(meta_leaf$sample)
res_leaf <- res[as.character(res$sample_id) %in% sample_leaf,]
meta_leaf <- dplyr::mutate(meta_leaf, path=as.character(res_leaf$kal_results))
meta_leaf
```

Create data matrix with tpm

```{r}
file <- "abundance.tsv"

full_file <- paste(res_leaf$kal_results, file, sep="/")

ab_list <- list()

for(i in 1:length(full_file)) {
  sample <- as.character(res_leaf$sample_id[i])
  
  ab <- read.table(full_file[i], header=T, sep="\t")
  #do.call("<-", list(res_leaf$sample_id[i], ab)) # assign each abundance file to its SRR
  
  #create the gene column
  transcripts <- as.character(ab$target_id)
  transcript_split <- strsplit(transcripts, "_")
  transcripts <- as.data.frame(do.call(rbind,transcript_split))
  colnames(transcripts) <- c("gene", "transcript_id")
  ab$gene <- transcripts$gene
  
  # get the gene counts
  g <- ab %>% 
    group_by(gene) %>%
    summarise(est_counts_sum=sum(tpm))
  #do.call("<-", list(genedf, g))
  g <- data.frame(g)
  rownames(g) <- g$gene
  g$gene <- NULL
  colnames(g) <- sample
  ab_list[[sample]] <- g
}

gene_tpm <- do.call(cbind, ab_list)

```

Create data matrix with estimated counts:

```{r}
ab_list <- list()

for(i in 1:length(full_file)) {
  sample <- as.character(res_leaf$sample_id[i])
  
  ab <- read.table(full_file[i], header=T, sep="\t")
  #do.call("<-", list(res_leaf$sample_id[i], ab)) # assign each abundance file to its SRR
  
  #create the gene column
  transcripts <- as.character(ab$target_id)
  transcript_split <- strsplit(transcripts, "_")
  transcripts <- as.data.frame(do.call(rbind,transcript_split))
  colnames(transcripts) <- c("gene", "transcript_id")
  ab$gene <- transcripts$gene
  
  # get the gene counts
  g <- ab %>% 
    group_by(gene) %>%
    summarise(est_counts_sum=sum(est_counts))
  #do.call("<-", list(genedf, g))
  g <- data.frame(g)
  rownames(g) <- g$gene
  g$gene <- NULL
  colnames(g) <- sample
  ab_list[[sample]] <- g
}

gene_counts <- do.call(cbind, ab_list)
```

Long data frames:

```{r}
rm(g, i)
gene_count_long <- gene_counts
gene_count_long$gene <- rownames(gene_counts)
gene_count_long <- reshape2::melt(gene_count_long)
colnames(gene_count_long) <- c("gene", "sample", "est_count")

meta_leaf2 <- meta_leaf %>% select(-path)
gene_count_long <- gene_count_long %>% left_join(meta_leaf2, by = "sample")

gene_tpm_long <- gene_tpm
gene_tpm_long$gene <- rownames(gene_tpm)
gene_tpm_long <- reshape2::melt(gene_tpm_long)
colnames(gene_tpm_long) <- c("gene", "sample", "tpm")
gene_tpm_long <- gene_tpm_long %>% left_join(meta_leaf2, by = "sample")
```

Reorder sample for plotting

```{r}
gene_tpm_long$sample <- as.factor(gene_tpm_long$sample)
gene_tpm_long$sample <- factor(gene_tpm_long$sample, 
                               levels = c("SRR1586888", "SRR1586889", "SRR1586766", "SRR1586767", "SRR1586768","SRR1586898","SRR1586899"))
```


# Analysis

## Design matrices

Full model

```{r}
meta_leaf$cultivar <- factor(meta_leaf$cultivar, levels=c("B73_TIL01", "B73", "TIL01"))
meta_leaf
full_leaf_model <- model.matrix(~cultivar, meta_leaf)
full_leaf_model
```

Make the reduced models:

```{r}
meta_leaf$B73 <- meta_leaf$cultivar == "B73"
meta_leaf$TIL01 <- meta_leaf$cultivar == "TIL01"
meta_leaf
```


Null model for difference between cross and B73

```{r}
reduced_TIL01 <- model.matrix(~1 + TIL01, meta_leaf)
reduced_TIL01
```

Null model for difference between cross and TIL01

```{r}
reduced_B73 <- model.matrix(~1+B73, meta_leaf)
reduced_B73
```

## TPM

#### Create DGE object

Possibly need a filtering step here.

```{r}
library(edgeR)
tpm_dge <- DGEList(counts = gene_tpm, group = as.factor(meta_leaf$sample))
tpm_dge <- calcNormFactors(tpm_dge)
```

**should I filter out for genes that are unexpressed across the board?**

#### voom/limma

Voom on full model

```{r}
tpm_v <- voom(tpm_dge, design = full_leaf_model, plot=TRUE)
```

Fit using limma

#### contrast for B73

Dropping the `cultivarB73` column of the design matrix.
Results from this section will show F1 crosses that are differentially expressed from B73, but not necessarily from TIL01.

```{r}
fit <- lmFit(tpm_v, design = full_leaf_model)
fit <- eBayes(fit)
tpm_fit_B73 <- topTable(fit, n=nrow(tpm_v), confint = TRUE, coef = "cultivarB73")
tpm_fit_B73$gene <- rownames(tpm_fit_B73)
tpm_B73_top <- tpm_fit_B73 %>% filter(adj.P.Val <= 0.01) %>% arrange(P.Value)
dim(tpm_B73_top)
head(tpm_B73_top)
```

Check out pvalues

```{r}
hist(tpm_fit_B73$P.Value, breaks= 100)
hist(p.adjust(tpm_fit_B73$P.Value, method = "BH"), breaks=100)
hist(tpm_fit_B73$adj.P.Val, breaks = 100)
```


```{r}
gene_tpm_long %>% filter(gene == tpm_B73_top[1,]$gene) %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle(tpm_B73_top[1,]$gene)
gene_tpm_long %>% filter(gene == tpm_B73_top[2,]$gene) %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle(tpm_B73_top[2,]$gene)
gene_tpm_long %>% filter(gene == tpm_B73_top[500,]$gene) %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle(tpm_B73_top[500,]$gene)
```

#### Contrast for TIL01

Drop the cultivarTIL01 from the design matrix. Anything that is significant prefers the alternative full model. Meaning that the variable TIL01 explains the result better.
Results from this section are differentially expressed from TIL01, but not necessarily differentially expressed from B73.

```{r}
tpm_fit_TIL01 <- topTable(fit, n=nrow(tpm_v), confint = TRUE, coef = "cultivarTIL01") 
tpm_fit_TIL01$gene <- rownames(tpm_fit_TIL01)
tpm_TIL01_top <- tpm_fit_TIL01 %>% filter(adj.P.Val <= 0.01) %>% arrange(P.Value)
dim(tpm_TIL01_top)
head(tpm_TIL01_top)
```

Check out pvalues

```{r}
hist(tpm_fit_TIL01$P.Value, breaks= 100)
hist(p.adjust(tpm_fit_TIL01$P.Value, method = "BH"), breaks=100)
```

what is that weird peak?

```{r}
tpm_fit_TIL01[which(tpm_fit_TIL01$P.Value > 0.7 & tpm_fit_TIL01 < 0.8),]
hist(tpm_fit_TIL01[which(tpm_fit_TIL01$P.Value > 0.7 & tpm_fit_TIL01 < 0.8),]$P.Value)
```

Zm00001d052931
```{r}
gene_tpm_long %>% filter(gene == "Zm00001d052931") %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle("Zm00001d052931")
```


```{r}
gene_tpm_long %>% filter(gene == tpm_TIL01_top[1,]$gene) %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle(tpm_TIL01_top[1,]$gene)
gene_tpm_long %>% filter(gene == tpm_TIL01_top[2,]$gene) %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle(tpm_TIL01_top[2,]$gene)
gene_tpm_long %>% filter(gene == tpm_TIL01_top[500,]$gene) %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle(tpm_TIL01_top[500,]$gene)
```

#### merge contrast results

```{r}
tpm_results <- merge(tpm_B73_top, tpm_TIL01_top, by="gene")
dim(tpm_results)
```

What percent of genes is that?

```{r}
nrow(tpm_results)/nrow(gene_tpm)
```

Plot the logFC

```{r}
tpm_results %>% ggplot() + geom_point(aes(x=logFC.x, y=logFC.y)) + 
  xlab("Log Fold Change (B73/F1)") + 
  ylab("Log Fold Change (TIL01/F1)")
```

The logFC is the difference between a parent and the hybrid. 


| Axes | sign | meaning |
| - | - | - |
| x | + | B73 > F1 |
| x | - | F1 > B73 |
|y | + | TIL01 > F1 |
| y | - | F1 > FIL01 |


```{r}
lfc <- dplyr::select(tpm_results, gene, logFC.x, logFC.y, P.Value.x, P.Value.y, adj.P.Val.x, adj.P.Val.y) %>% arrange(logFC.x)

head(lfc)
```



```{r}
gene_tpm_long %>% filter(gene == "Zm00001d024522") %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle("Zm00001d024522")
gene_tpm_long %>% filter(gene == "Zm00001d024522")
lfc %>% filter(gene == "Zm00001d024522")
```

Zm00001d033389
```{r}
gene_tpm_long %>% filter(gene == "Zm00001d033389") %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle("Zm00001d033389")
gene_tpm_long %>% filter(gene == "Zm00001d033389")
lfc %>% filter(gene == "Zm00001d033389")
```

Zm00001d030028
```{r}
gene_tpm_long %>% filter(gene == "Zm00001d030028") %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle("Zm00001d030028") +
  ylab("transcripts per million")
gene_tpm_long %>% filter(gene == "Zm00001d030028")
lfc %>% filter(gene == "Zm00001d030028")
```

Opposing logFC

Zm00001d005583
```{r}
gene_tpm_long %>% filter(gene == "Zm00001d005583") %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle("Zm00001d005583")
lfc %>% filter(gene == "Zm00001d005583")
```

Zm00001d050635
```{r}
gene_tpm_long %>% filter(gene == "Zm00001d050635") %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle("Zm00001d050635")
lfc %>% filter(gene == "Zm00001d050635")
```

Zm00001d028541
```{r}
gene_tpm_long %>% filter(gene == "Zm00001d028541") %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle("Zm00001d028541")
lfc %>% filter(gene == "Zm00001d028541")
```

Zm00001d052340
```{r}
gene_tpm_long %>% filter(gene == "Zm00001d052340") %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle("Zm00001d052340")
lfc %>% filter(gene == "Zm00001d052340")
```

Zm00001d017699
```{r}
gene_tpm_long %>% filter(gene == "Zm00001d017699") %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle("Zm00001d052340")
lfc %>% filter(gene == "Zm00001d017699")
```

### Number of genes overdominant

```{r}
overDom <- tpm_results %>% filter(logFC.x > 0 & logFC.y > 0)
dim(overDom)
```

```{r}
underDom <- tpm_results %>% filter(logFC.x < 0 & logFC.y < 0)
dim(underDom)
```

```{r}
sum(nrow(overDom), nrow(underDom))/nrow(tpm_results)
```


## Design matrix for F1

```{r}
meta_leaf$F1 <- meta_leaf$cultivar == "B73_TIL01"
meta_leaf$cultivar <- factor(meta_leaf$cultivar, levels=c("B73", "TIL01", "B73_TIL01"))
f1_model <- model.matrix(~1+F1, meta_leaf)
```

voom

```{r}
tpm_f1_v <- voom(tpm_dge, design = f1_model, plot=TRUE)

fit_f1 <- lmFit(tpm_f1_v, design = full_leaf_model)
fit_f1 <- eBayes(fit_f1)
tpm_fit_f1 <- topTable(fit_f1, n=nrow(tpm_f1_v), confint = TRUE)
tpm_fit_f1$gene <- rownames(tpm_fit_f1)
tpm_f1_top <- tpm_fit_f1 %>% filter(adj.P.Val <= 0.05) %>% arrange(P.Value)
dim(tpm_f1_top)
```

```{r}
head(tpm_f1_top)
```

```{r}
gene_tpm_long %>% filter(gene == "Zm00001d003796") %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity")
gene_tpm_long %>% filter(gene == "Zm00001d024522") %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity")
gene_tpm_long %>% filter(gene == "Zm00001d011183") %>% ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity")
```

```{r}
tpm_f1_top %>% ggplot() + geom_point(aes(x=cultivarB73, y=cultivarTIL01))
```

Not in previous results

```{r}
dunno <- tpm_f1_top[!(tpm_f1_top$gene %in% tpm_results$gene),]
dim(dunno)
tpm_f1_top %>% ggplot() + 
  geom_point(aes(x=cultivarB73, y=cultivarTIL01)) + 
  geom_point(data = dunno, aes(x=cultivarB73, y=cultivarTIL01), colour="red") +
  geom_point(data = tpm_results, aes(x=logFC.x, y=logFC.y), colour="blue")
```

```{r}
gene_tpm_long %>% 
  filter(gene == dunno[1,]$gene)%>% 
  ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") + ggtitle(dunno[1,]$gene)
gene_tpm_long %>% 
  filter(gene == dunno[2,]$gene)%>% 
  ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") + ggtitle(dunno[2,]$gene)
gene_tpm_long %>% 
  filter(gene == dunno[500,]$gene)%>% 
  ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") + ggtitle(dunno[500,]$gene)
gene_tpm_long %>% 
  filter(gene == dunno[5000,]$gene)%>% 
  ggplot() + geom_bar(aes(x=sample, y=tpm, fill=cultivar), stat = "identity") + ggtitle(dunno[5000,]$gene)
```

How many genes from this dataset are under and over dominant?

```{r}
f1_overDom <- tpm_f1_top %>% filter(cultivarB73 < 0 & cultivarTIL01 < 0)
f1_underDom <- tpm_f1_top %>% filter(cultivarB73 > 0 & cultivarTIL01 > 0)
dim(f1_overDom); dim(f1_underDom)
```

Quite a few more genes...

```{r}
ggplot() + geom_point(data=f1_overDom, aes(x=cultivarB73, y=cultivarTIL01)) +
  geom_point(data=f1_underDom, aes(x=cultivarB73, y=cultivarTIL01))
```


# Cross reference my significant genes with Lemmon et al regulatory classifications

Read in Lemmon et al. Supp data. This has how they classified genes in terms of their regulatory role.

It has two header columns, so I need to get fancy.

```{r}
header1 <- scan("data/Lemmon/SupplementalDataset1_v3.csv", nlines = 1, what = character(), sep = ",")
header2 <- scan("data/Lemmon/SupplementalDataset1_v3.csv", skip=1, nlines=1, what = character(), sep=",")
lemmon <- read.delim("data/Lemmon/SupplementalDataset1_v3.txt", skip=2, sep="\t", header=F)
colnames(lemmon) <- paste(header1, header2, sep="_")
lemmon_leaf <- lemmon %>% select(gene="_Gene", "Leaf_HybridMaize","Leaf_HybridTeosinte",
                                 "Leaf_ParentMaize","Leaf_ParentTeosinte" ,"Leaf_Hybrid.binom.p", "Leaf_Parent.binom.p", 
                                 "Leaf_trans.p","Leaf_Reg.Cat","Leaf_Crosses","Leaf_CCTlistLong","Leaf_R2_Maize",
                                 "Leaf_R2_Teosinte")
```

Lemmon used AGPv2, and so I  need to convert between AGPv2 gene ids to AGPv4, which I used.

genome conversion data
```{r}
genome_convert <- read.delim("~/Box Sync/Projects/genomes/maize/MaizeGDB_v3_v4.genes.csv", sep=",", header=T, na.strings = "na")
genome_convert <- genome_convert %>% select(v3.gene.ID, v4.gene.ID = v4.gene.ID..if.present.)
genome_special <- read.delim("~/Box Sync/Projects/genomes/maize/updated_models_v3.txt", header=T)
```

convert Lemmon gene column

```{r}
dim(lemmon_leaf)
lemmon_convert <- genome_convert[genome_convert$v3.gene.ID %in% lemmon_leaf$gene,]
lemmon_leaf$gene.v4 <- 
```

